### ISSUE TITLE
astimezone() fails on Windows for pre-epoch times

### ISSUE THREAD

**Issue by 05baa077-d688-4697-affd-f7a78c25afa4** [python/cpython#80940]
BPO | [36759](https://bugs.python.org/issue36759)
--- | :---
Nosy | @abalkin, @pganssle, @Windsooon, @jonnyhsu

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-04-30.11:36:25.986>
labels = ['3.7', '3.8', 'type-bug', 'library', '3.9']
title = 'astimezone() fails on Windows for pre-epoch times'
updated_at = <Date 2020-03-24.08:53:35.045>
user = 'https://bugs.python.org/Snidhi'
```

bugs.python.org fields:
```python
activity = <Date 2020-03-24.08:53:35.045>
actor = 'Ard Kuijpers'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-04-30.11:36:25.986>
creator = 'Snidhi'
dependencies = []
files = []
hgrepos = []
issue_num = 36759
keywords = []
message_count = 10.0
messages = ['341149', '341152', '341157', '341236', '342866', '364801', '364866', '364907', '364918', '364921']
nosy_count = 7.0
nosy_names = ['belopolsky', 'SilentGhost', 'p-ganssle', 'Snidhi', 'Windson Yang', 'Jonathan Hsu', 'Ard Kuijpers']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'test needed'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue36759'
versions = ['Python 3.7', 'Python 3.8', 'Python 3.9']
```

</p></details>

**Comment by bariod** [comment]
For some reason, for post-epoch values, this works till the year 3001, not 2038:
```py
from datetime import datetime
datetime(3001, 1, 19, 7, 59, 59, 999999).astimezone()  # still works
datetime(3001, 1, 19, 8, 0, 0, 0).astimezone() # fails with 'OSError: [Errno 22] Invalid argument'
```
*run with py 3.9.6*

**Comment by dnowacki-usgs** [comment]
> [PEP-615](https://peps.python.org/615) proposes to include the IANA tz database, which would negate the need for a system call. Should we wait for this PEP before fixing this issue? Thoughts?

FWIW, PEP-615 was accepted and became part of Python 3.9 in May 2020, so perhaps this approach is viable now?

**Comment by pganssle** [comment]
> FWIW, [PEP-615](https://peps.python.org/615) was accepted and became part of Python 3.9 in May 2020, so perhaps this approach is viable now?

PEP 615 didn't really change anything with respect to this. The issue is that [naïve times are local times](https://blog.ganssle.io/articles/2022/04/naive-local-datetimes.html), and need to be resolved with the system's time zone handling mechanisms. The linked article goes into detail about why it is not sufficient for all use cases to attach a tzinfo that corresponds to "local time" (though individuals are free to do that if they have one of the many very common situations where these concerns don't matter).

This one may be a fundamental limitation of the Windows API, though I doubt it's a major issue, since time zone handling doesn't make nearly as much sense the further back (or forward!) in time you go. It seems unusual for someone to actually care about the answer to the question, "What time zone does this computer think should be attached to a date in local time in 1965?" for example.

**Comment by dnowacki-usgs** [comment]
Thanks for the insight here re: the Windows API.

> It seems unusual for someone to actually care about the answer to the question, "What time zone does this computer think should be attached to a date in local time in 1965?" for example.

In my specific case, we have decades of environmental observations, often made pre-1970, tagged in a database with times in the local time zone for the observation. When retrieving these data we wish to unify them all to GMT/UTC.

**Comment by pganssle** [comment]
> In my specific case, we have decades of environmental observations, often made pre-1970, tagged in a database with times in the local time zone for the observation. When retrieving these data we wish to unify them all to GMT/UTC.

Yes, but these observations were not made on the computer you are running the code from, and even if they were, there's nothing that says they were using the set of time zone rules that _currently_ applies on the computer you are running the code from. For code like that you need to either know the UTC time or the local time zone rules that were in effect at the time (or possibly both), and the built-in support for local time would be at best coincidentally useful.
